language/move-lang/src/cfgir/cfg.rs

```diff
@@ -70,31 +87,145 @@ impl<'a> BlockCFG<'a> {
         (cfg, errors)
     }
 
-    pub fn successors(&self, label: &Label) -> &BTreeSet<Label> {
-        self.successor_map.get(label).unwrap()
+    pub fn block(&self, label: Label) -> &BasicBlock {
+        self.blocks.get(&label).unwrap()
+    }
+
+    pub fn block_mut(&mut self, label: Label) -> &mut BasicBlock {
+        self.blocks.get_mut(&label).unwrap()
+    }
+}
+
+impl<'a> CFG for BlockCFG<'a> {
+    fn successors(&self, label: Label) -> &BTreeSet<Label> {
+        self.successor_map.get(&label).unwrap()
+    }
+
+    fn predecessors(&self, label: Label) -> &BTreeSet<Label> {
+        self.predecessor_map.get(&label).unwrap()
+    }
+
+    fn commands<'s>(&'s self, label: Label) -> Box<dyn Iterator<Item = &'s Command> + 's> {
+        Box::new(self.block(label).iter())
+    }
+
+    fn commands_mut<'s>(
+        &'s mut self,
+        label: Label,
+    ) -> Box<dyn Iterator<Item = &'s mut Command> + 's> {
+        Box::new(self.block_mut(label).iter_mut())
+    }
+
+    fn num_blocks(&self) -> usize {
+        self.blocks.len()
+    }
+
+    fn start_block(&self) -> Label {
+        self.start
+    }
+}
+
+//**************************************************************************************************
+// BlockCFG
+//**************************************************************************************************
+
+#[derive(Debug)]
+pub struct ReverseBlockCFG<'a> {
+    start: Label,
+    blocks: &'a mut Blocks,
+    successor_map: BTreeMap<Label, BTreeSet<Label>>,
+    predecessor_map: BTreeMap<Label, BTreeSet<Label>>,
+}
+
+impl<'a> ReverseBlockCFG<'a> {
+    pub fn new(forward_cfg: &'a mut BlockCFG, infinite_loop_starts: BTreeSet<Label>) -> Self {
+        let blocks: &'a mut Blocks = &mut forward_cfg.blocks;
+        let forward_successors = &forward_cfg.successor_map;
+        let forward_predecessor = &forward_cfg.predecessor_map;
+        let end_blocks = {
+            let mut end_blocks = BTreeSet::new();
+            for (lbl, successors) in forward_successors {
+                let loop_start_successors = successors
+                    .iter()
+                    .filter(|l| infinite_loop_starts.contains(l));
+                for loop_start_successor in loop_start_successors {
+                    if lbl >= loop_start_successor {
+                        end_blocks.insert(*lbl);
+                    }
+                }
+            }
+            for (lbl, block) in blocks.iter() {
+                let last_cmd = block.back().unwrap();
+                if last_cmd.value.is_divergent() {
+                    end_blocks.insert(*lbl);
+                }
+            }
+            end_blocks
+        };
+
+        let mut successor_map = forward_predecessor.clone();
```

Comment:
Nit: might be less confusing to keep the usual interpretation of predecessors and successors for as long as possible (e.g., `predecessor_map.instert(terminal, end_blocks)` below instead), then do `successor_map: predecessor_map, predecessor_map: succesor_map` when you actually build the struct.

Reply:
Done